Library/OcDeviceMiscLib/SetResizableBar.c

```diff
@@ -309,8 +377,241 @@ SetResizableBarOnDevice (
+  IN UINTN                            PciAddress,
+  IN PCI_BAR_SIZE                     Size,
+  IN BOOLEAN                          Increase
+  )
+{
+  EFI_STATUS                                               Status;
+  UINT32                                                   ResizableBarOffset;
+  PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_ENTRY    Entries[PCI_MAX_BAR];
+  PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_CONTROL  ResizableBarControl;
+  UINT32                                                   Offset;
+  UINT32                                                   Index;
+  UINT32                                                   ResizableBarNumber;
+  UINT64                                                   Capabilities;
+  UINT64                                                   NewCapabilities;
+  UINT32                                                   OldBar[PCI_MAX_BAR];
+  UINT32                                                   NewBar[PCI_MAX_BAR];
+  INTN                                                     Bit;
+  BOOLEAN                                                  ChangedBars;
+
+  ChangedBars = FALSE;
+
+  Status = LocatePciCapabilityRbIo (
+             PciRootBridgeIo,
+             PciAddress,
+             PCI_EXPRESS_EXTENDED_CAPABILITY_RESIZABLE_BAR_ID,
+             &ResizableBarOffset
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCDM: RBAR is unsupported by device - %r\n", Status));
+    return EFI_UNSUPPORTED;
+  }
+
+  ResizableBarControl.Uint32 = 0;
+  Offset                     = ResizableBarOffset + sizeof (PCI_EXPRESS_EXTENDED_CAPABILITIES_HEADER)
+                               + sizeof (PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_CAPABILITY);
+  Status = PciRootBridgeIo->Pci.Read (
+                                  PciRootBridgeIo,
+                                  EfiPciWidthUint8,
+                                  PciAddrOffset (PciAddress, Offset),
+                                  sizeof (PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_CONTROL),
+                                  &ResizableBarControl
+                                  );
+
+  DEBUG ((
+    DEBUG_INFO,
+    "OCDM: RBAR control is %X, total %u - %r\n",
+    ResizableBarControl.Uint32,
+    MIN (ResizableBarControl.Bits.ResizableBarNumber, PCI_MAX_BAR),
+    Status
+    ));
+
+  if (EFI_ERROR (Status)) {
+    return EFI_UNSUPPORTED;
+  }
+
+  ResizableBarNumber = MIN (ResizableBarControl.Bits.ResizableBarNumber, PCI_MAX_BAR);
+
+  Status = PciRootBridgeIo->Pci.Read (
+                                  PciRootBridgeIo,
+                                  EfiPciWidthUint8,
+                                  PciAddrOffset (PciAddress, ResizableBarOffset + sizeof (PCI_EXPRESS_EXTENDED_CAPABILITIES_HEADER)),
+                                  sizeof (PCI_EXPRESS_EXTENDED_CAPABILITIES_RESIZABLE_BAR_ENTRY) * ResizableBarNumber,
+                                  (VOID *)Entries
```

Comment:
Which compiler requires these kinds of cast?

Reply:
Also was taken from PciIo version but will try to fix it